ir/src/codegen/codegen.rs

```diff
@@ -0,0 +1,442 @@
+    locals_map: &SecondaryMap<Value, u16>,
+    insts: &mut Vec<MASMNode>,
+) {
+    // Push arguments onto the stack
+    let args = source.arguments(pool);
+    for i in 0..args.len() {
+        let v = args[i];
+        insts.push(masm_instruction_as_node(MASMInstruction::LocStore(
+            locals_map[v],
+        )));
+    }
+
+    // Perform the op
+    // TODO: We are currently assuming that all operations are 32-bit. This will not be a valid assumption in general.
+    let op = source.opcode();
+    let inst = match source {
+        IRInstruction::BinaryOp(BinaryOp { ref overflow, .. }) => codegen_binary_op(op, overflow),
+        IRInstruction::BinaryOpImm(BinaryOpImm {
+            ref overflow,
+            ref imm,
+            ..
+        }) => codegen_binary_op_imm(op, overflow, immediate_as_u32(imm)),
+        IRInstruction::UnaryOp(_) => codegen_unary_op(op),
+        IRInstruction::UnaryOpImm(UnaryOpImm { ref imm, .. }) => {
+            codegen_unary_op_imm(op, immediate_as_u32(imm))
+        }
+        IRInstruction::Call(_) => codegen_call(op),
+        IRInstruction::Br(_) => codegen_br(op),
+        IRInstruction::CondBr(_) => codegen_cond_br(op),
+        IRInstruction::Switch(_) => codegen_switch(op),
+        IRInstruction::Ret(_) => codegen_ret(op),
+        IRInstruction::MemCpy(_) => codegen_mem_cpy(op),
+        IRInstruction::PrimOp(_) => codegen_prim_op(op),
+        IRInstruction::PrimOpImm(PrimOpImm { ref imm, .. }) => {
+            codegen_prim_op_imm(op, immediate_as_u32(imm))
+        }
+        IRInstruction::Test(_) => codegen_test(op),
+        IRInstruction::InlineAsm(_) => codegen_inline_asm(op),
+    };
+    insts.push(inst);
+}
+
+fn codegen_function(source: &Function) -> MASMProcedureAst {
+    let dfg = &source.dfg;
+    //Determine the number of locals. We assign
+    // - a local for each value in the dfg
+    // - a local for each parameter
+    let locals_count = dfg.values.len();
+    let mut locals_map: SecondaryMap<Value, u16> = SecondaryMap::with_capacity(locals_count);
+    let mut i = 0;
+    for v in dfg.values.keys() {
+        locals_map[v] = i;
+        i += 1;
+    }
+
+    let mut compiled_instructions: Vec<MASMNode> = Vec::new();
+
+    //Parameters are initially stored on the stack, but we need to read them from local memory, so store parameters as locals
+    let entry_block = dfg
+        .entry_block()
+        .expect("No entry block found for function {source}");
+    for v in dfg.block_params(entry_block) {
+        compiled_instructions.push(masm_instruction_as_node(MASMInstruction::LocStore(
+            locals_map[*v],
+        )));
+    }
+
+    //TODO: Handle branches properly - we currently assume no branching = only one block.
+
+    debug_assert!(dfg.num_blocks() == 1, "too many blocks");
+    for inst in dfg.block_insts(entry_block) {
+        let i = &dfg.insts[inst];
+        codegen_instruction(
+            &i.item,
+            &dfg.value_lists,
```

Comment:
I would pass the entire `DataFlowGraph` in here, as it is the only way to get instruction results, which it appears are currently not being handled.

I don't think this will generate working code at the moment - the reason being that while the result of an expression is on the Miden operand stack, that value is not being stored into the local corresponding to the instruction result(s) in Miden IR. Thus, the next instruction that tries to use that local will get garbage (likely `0` in simple programs, as the memory hasn't been allocated into yet) instead of the expected value.

For example, consider the following function in Miden IR:

```
pub fn multiply_and_add(i32, i32, i32): i32 {
entry(a: i32, b: i32, c: i32):
    v0 = mul a, b
    v1 = add v0, c
    ret v1
}
```

The current lowering does not store the result of `mul a, b` in `v0`, so the subsequent `add v0, c` is non-deterministic/incorrect. Similarly, the `add` doesn't store its result in `v1`, so the `ret` will just return whatever value is in the memory allocated for `v1`.

Reply:
`store` operation added. As you said, the dataflow graph now needs to be passed to `codegen_instruction`.